pkg/controller/advancedcronjob/advancedcronjob_imagelistpulljob_controller.go

```
@@ -0,0 +1,393 @@
+	if err := r.updateAdvancedJobStatus(req, &advancedCronJob); err != nil {
+		klog.ErrorS(err, "Unable to update AdvancedCronJob status", "advancedCronJob", req)
+		return ctrl.Result{}, err
+	}
+
+	/*
+		Once we've updated our status, we can move on to ensuring that the status of
+		the world matches what we want in our spec.
+		### 3: Clean up old jobs according to the history limit
+		First, we'll try to clean up old jobs, so that we don't leave too many lying
+		around.
+	*/
+
+	// NB: deleting these is "best effort" -- if we fail on a particular one,
+	// we won't requeue just to finish the deleting.
+	if advancedCronJob.Spec.FailedJobsHistoryLimit != nil {
+		sort.Slice(failedJobs, func(i, j int) bool {
+			if failedJobs[i].Status.StartTime == nil {
+				return failedJobs[j].Status.StartTime != nil
+			}
+			return failedJobs[i].Status.StartTime.Before(failedJobs[j].Status.StartTime)
+		})
+		for i, job := range failedJobs {
+			if int32(i) >= int32(len(failedJobs))-*advancedCronJob.Spec.FailedJobsHistoryLimit {
+				break
+			}
+
+			if err := r.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+				klog.ErrorS(err, "Unable to delete old failed ImageListPullJob", "job", klog.KObj(job), "advancedCronJob", req)
+			} else {
+				klog.InfoS("Deleted old failed ImageListPullJob", "job", klog.KObj(job), "advancedCronJob", req)
+			}
+		}
+	}
+
+	if advancedCronJob.Spec.SuccessfulJobsHistoryLimit != nil {
+		sort.Slice(successfulJobs, func(i, j int) bool {
+			if successfulJobs[i].Status.StartTime == nil {
+				return successfulJobs[j].Status.StartTime != nil
+			}
+			return successfulJobs[i].Status.StartTime.Before(successfulJobs[j].Status.StartTime)
+		})
+		for i, job := range successfulJobs {
+			if int32(i) >= int32(len(successfulJobs))-*advancedCronJob.Spec.SuccessfulJobsHistoryLimit {
+				break
+			}
+
+			if err := r.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
```

plz ignore error not found